### ISSUE TITLE
New router behavior around middleware breaks redirects from getInitialProps on CSR

### ISSUE THREAD

**Issue by ParallelUniv3rse** [vercel/next.js#39126]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

    Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 21.4.0: Fri Mar 18 00:47:26 PDT 2022; root:xnu-8020.101.4~15/RELEASE_ARM64_T8101
    Binaries:
      Node: 16.10.0
      npm: 7.24.0
      Yarn: 1.22.15
      pnpm: 6.11.0
    Relevant packages:
      next: 12.2.0
      eslint-config-next: N/A
      react: 17.0.2
      react-dom: 17.0.2

### Describe the Bug

The redirect happens as a rewrite instead of a redirect. The correct redirected page displays, but the URL stays as the page we've initially navigated to. 

### Expected Behavior

A normal redirect happens.

### Link to reproduction

https://codesandbox.io/s/hungry-glitter-isub6z?file=/pages/content.tsx

### To Reproduce

Redirect the user inside a `getInitialProps` call while also using middleware in your project.

A sample redirection logic can be seen in the codesandbox in `/routing.ts`

**Comment by codynova** [comment]
This also occurs with `redirects` defined in `next.config.js`, see reproductions for Next 12.2 and Next 12.3 here https://github.com/codynova/next-middleware-redirects

**Comment by ParallelUniv3rse** [comment]
This issue still persists on the new v14.0.0. We still cannot upgrade our apps without a major rewrite.

**Comment by ParallelUniv3rse** [comment]
For anyone concerned with this issue - I have investigated further and found that this issue can be worked around in user-land. 

TL;DR: The problem is prevented by not redirecting on a data request in the server-side part of a custom redirect function. Example below: 
```
if (res) {
    const isDataReq = ctx.req?.headers["x-nextjs-data"]; // checking for a middleware data request prevents this issue

    if (!isDataReq) {
      // Server redirect
      res.writeHead(push ? 307 : 302, {
        Location: as || location
      });
    }

    res.end();
  } else {
  ...
 ```
 
To elaborate - the issue is caused by a redirected url being returned as an effect from the `fetchData` request in the `withMiddlewareEffects` function in router (https://github.com/vercel/next.js/blob/6c6ac700b221f2018a03b05c16224055b15629f4/packages/next/src/shared/lib/router/router.ts#L327).  The `fetchData` request returns with the redirected url in the `x-nextjs-matched-path` header if a redirect happens in getInitialProps, when performing a data request.
In case of my demo from the reproduction, the request url is `"/_next/data/development/content.json"`. 
Such request can be regognized in the getInitialProps function by the incoming `x-nextjs-data: "1"` header and then not redirecting in such case.

It would be great to have this handled internally in next, so userland does not have to rely on fixing a common pattern used for routing in nextjs < 10.
